Framework/CurveFitting/src/Functions/PseudoVoigt.cpp

```diff
@@ -17,83 +18,362 @@ namespace Functions {
+ */
+void PseudoVoigt::update_set_history(size_t set_index) {
+  if (set_index > 3)
+    throw std::runtime_error("Parameter set up history index out of boundary");
+
+  size_t prev_distance = m_set_history_distances[set_index];
+  for (size_t i = 0; i < 4; ++i) {
+    // no operation on any never-set or out-dated index
+    if (m_set_history_distances[i] >= 3)
+      continue;
+
+    // only increase the historoy distance by 1 in case
+    // the other parameters with those with distance less than the previous one
+    // i.e., out of date or never been set
+    if (m_set_history_distances[i] < prev_distance) {
+      m_set_history_distances[i] += 1;
+      // if out of range, don't worry about it
+    }
+  }
+
+  // set the current one to 0 (most recent)
+  m_set_history_distances[set_index] = 0;
+}
+
+/** As a parameter (height, width, intensity and mixing) being set (via
+ * setParameter)
+ * find out whethere are are enough parameters that have been specified
+ * @brief PseudoVoigt::get_parameter_to_calculate_from_set
+ * @return :: index of the parameter to calculate if there is
+ */
+size_t PseudoVoigt::get_parameter_to_calculate_from_set() {
+  // go through setting-history-distance
+  size_t index_largest_distance{0};
+  size_t largest_distance{0};
+  size_t num_been_set{0};
+
+  // std::cout << "distances size = " << m_set_history_distances.size() << "\n";
+  for (size_t i = 0; i < m_set_history_distances.size(); ++i) {
+    std::cout << "distance[" << i << "] = " << m_set_history_distances[i]
+              << "\n";
+    if (m_set_history_distances[i] >= largest_distance) {
+      // for those to set for: any number counts
+      index_largest_distance = i;
+      largest_distance = m_set_history_distances[i];
+    }
+    // only those been set and not out of date counted
+    if (m_set_history_distances[i] < 3) {
+      ++num_been_set;
+    }
+  }
+
+  // signal for no enough index been set up
+  if (num_been_set < 3)
+    index_largest_distance = 128;
+
+  return index_largest_distance;
+}
+
+/** calculate effective parameter: height
+ * @return
+ */
+double PseudoVoigt::height() const {
+  double peak_intensity = getParameter("Intensity");
+  double gamma = getParameter("FWHM");
+  double eta = getParameter("Mixing");
+  const double height = 2. * peak_intensity *
+                        (1 + (sqrt(M_PI * M_LN2) - 1) * eta) / (M_PI * gamma);
+
+  return height;
+}
+
+/** set non-native parameter peak height
+ * @brief PseudoVoigt::setHeight
+ * @param h
+ */
+void PseudoVoigt::setHeight(const double h) {
+  // set height
+  m_height = h;
+  update_set_history(2);
+  std::cout << "PV: set height = " << m_height << "\n";
+  estimate_parameter_value();
+}
+
+/** set FWHM
+ * @param w
+ */
+void PseudoVoigt::setFwhm(const double w) { setParameter("FWHM", w, true); }
+
+/** a_G = 2/gamma * sqrt(ln2/pi)
+ * @param gamma :: FWHM
+ * @return
+ */
+double PseudoVoigt::cal_ag(const double gamma) const {
```

Comment:
Shouldn't be variable be named "sigma" as that is the greek letter?

Reply:
this is to set FHWM but not sigma